hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/upgrade/ReconLayoutVersionManager.java

```diff
@@ -0,0 +1,135 @@
+ * Recon service, and finalizes the layout features that need to be upgraded.
+ */
+public class ReconLayoutVersionManager {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(ReconLayoutVersionManager.class);
+
+  // Current Metadata Layout Version (MLV) of the Recon service
+  private int currentMLV;
+  // Schema Layout Version (SLV) of the Recon service
+  private final int currentSLV;
+  private final ReconSchemaVersionTableManager schemaVersionTableManager;
+
+  public ReconLayoutVersionManager(ReconSchemaVersionTableManager schemaVersionTableManager) {
+    this.schemaVersionTableManager = schemaVersionTableManager;
+    this.currentMLV = determineMLV();
+    this.currentSLV = determineSLV();
+  }
+
+  /**
+   * Determines the current Metadata Layout Version (MLV) from the version table.
+   * @return The current Metadata Layout Version (MLV).
+   */
+  private int determineMLV() {
+    return schemaVersionTableManager.getCurrentSchemaVersion();
+  }
+
+  /**
+   * Determines the Schema Layout Version (SLV) based on the latest feature version.
+   * @return The Schema Layout Version (SLV).
+   */
+  private int determineSLV() {
+    return Arrays.stream(ReconLayoutFeature.values())
+        .mapToInt(ReconLayoutFeature::getVersion)
+        .max()
+        .orElse(0); // Default to 0 if no features are defined
+  }
+
+  /**
+   * Finalizes the layout features that need to be upgraded, by executing the upgrade action for each
+   * feature that is registered for finalization.
+   */
+  public void finalizeLayoutFeatures() {
+    // Get features that need finalization, sorted by version
+    List<ReconLayoutFeature> featuresToFinalize = getRegisteredFeatures();
+
+    for (ReconLayoutFeature feature : featuresToFinalize) {
+      try {
+        // Execute the upgrade action for this feature
+        feature.getUpgradeAction().execute();
+        // Update the MLV in the database after successful execution
+        updateSchemaVersion(feature.getVersion());
+        LOG.info("Feature " + feature.getVersion() + " finalized successfully.");
+      } catch (Exception e) {
+        LOG.info("Failed to finalize feature " + feature.getVersion() + ": " + e.getMessage());
+        break;
+      }
+    }
+  }
+
+  /**
+   * Returns a list of ReconLayoutFeature objects that are registered for finalization.
+   * @return List of ReconLayoutFeature objects that are registered for finalization.
+   */
+  private List<ReconLayoutFeature> getRegisteredFeatures() {
+    List<ReconLayoutFeature> allFeatures =
+        Arrays.asList(ReconLayoutFeature.values());
+
+    LOG.info("Current MLV: {}. SLV: {}. Checking features for registration...", currentMLV, currentSLV);
+
+    List<ReconLayoutFeature> registeredFeatures = allFeatures.stream()
+        .filter(feature -> {
+          boolean shouldRegister =
+              feature.getVersion() > currentMLV && feature.getVersion() <= currentSLV;
```

Comment:
`feature.getVersion() <= currentSLV`, here comparing the feature version with less than current SLV, pls check if it is needed.

Reply:
Done!

Reply:
Done